### PR TITLE
profanity checking on text input

### PR DESCRIPTION
**Pull request by Rand0mLetters**
added profanity checking to challenge creation wizard
## Description

still need to make a text input with anti-profanity, working on that

## Related Issue

#274 

fine, I'll link the issue (again)
https://github.com/typehero/typehero/issues/274

## Motivation and Context

argh why is this form so long

[yes, I will link to the issue *again*](https://github.com/typehero/typehero/issues/274)

## How Has This Been Tested?

by typing lots and lots of swear words into the inputs

## Screenshots/Video (if applicable):
the "don't use profanity" at the bottom is a result of absolute positioning and local changes
image

### CODE REVIEW

**Comment by Rand0mLetters** [comment]
still wip, don't merge (I'll fix prettier and other issues tomorrow or something)

File: apps/web/src/components/wizard/index.tsx [hunk 1]
```diff
@@ -18,27 +18,34 @@ import { uploadChallenge } from './create.action';
...
+
 const testCaseRegex = new RegExp('(?:\n|^)s*(?:Equal|Extends|NotEqual|Expect)<');
 const createExploreCardSchema = z.object({
   difficulty: z.enum(['BEGINNER', 'EASY', 'MEDIUM', 'HARD', 'EXTREME']),
   name: z
     .string()
     .min(3, 'The name must be longer than 3 characters')
-    .max(30, 'The name must be shorter than 30 characters'),
+    .max(30, 'The name must be shorter than 30 characters')
+    .refine((str: string) => !containsProfanity(str), "The name must not contain bad words."),
```
Comment (bautistaaa) [inline apps/web/src/components/wizard/index.tsx:38]: we can probably just make a custom zod type so you dont have to repeat this

Comment (Rand0mLetters) [reply apps/web/src/components/wizard/index.tsx:38]: it's really only used here (I think), and I'm working on a textarea that prevents submission if there's profanity, so it might not be necessary

I'll see how it goes with the textarea

File: apps/web/src/utils/profanity.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import Filter from 'bad-words';
+import overrideWords from './bad-word-overrides.json';
+
+
+export function containsProfanity(str: string): boolean {
+    const f = new Filter();
```
Comment (bautistaaa) [inline apps/web/src/utils/profanity.ts:6]: can move this outside the function otherwise it constructs a new class every invocation

Comment (Rand0mLetters) [reply apps/web/src/utils/profanity.ts:6]: oh shoot, yeah I'll do that

File: apps/web/src/components/wizard/index.tsx (hunk 1 above)
Comment (bautistaaa) [reply apps/web/src/components/wizard/index.tsx:38]: this needs to be applied to any place youre allowed to use text. comments, solutions, challenge submissions

File: apps/web/src/app/[locale]/[username]/page.tsx [hunk 3]
```diff
@@ -11,8 +11,7 @@ export default function Page({ params }: Props) {
 }
 
 export async function generateMetadata({ params: { username } }: Props) {
-  // remove the %40 from the username
-  const name = username.substring(3);
+  const name = decodeURIComponent(username).substring(1);
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/[username]/page.tsx:14]: why decode just to remove?

File: apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx [hunk 4]
```diff
@@ -21,6 +22,7 @@ export function CommentInput({ mode, onCancel, onChange, value, placeholder, onS
   const { toast } = useToast();
   const [commentMode, setCommentMode] = useState<'editor' | 'preview'>('editor');
   const [isSubmitting, setIsSubmitting] = useState(false);
+  const [submittable, setSubmittable] = useState(true);
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx:25]: can't we just leverage zod here and change this into a form

File: apps/web/src/app/[locale]/[username]/page.tsx (hunk 3 above)
Comment (Rand0mLetters) [reply apps/web/src/app/[locale]/[username]/page.tsx:14]: some people hve spaces in their name (from seed data, @Alexandra Osinski) and the spaces are %20 which is ugly

File: apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx (hunk 4 above)
Comment (Rand0mLetters) [reply apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx:25]: probably a good idea but I haven't worked with zod before. I'll try to make it work

**Comment by Rand0mLetters** [comment]
added profanity checking to /settings
I think that is all the text input there is? @bautistaaa am I missing anything?
branch ready for merge (with review) I think

image

**Comment by Rand0mLetters** [comment]
I am missing something, will update the solution submission form rn

File: apps/og-image/utils/zodParams.ts [hunk 5]
```diff
@@ -52,15 +66,15 @@ export const fontParams = zodParams(
 
 export const challengeParam = zodParams(
   z.object({
-    title: truncatedWordSchema({ maxCharacters: 70 }),
-    description: truncatedWordSchema({ maxCharacters: 145 }),
-    username: z.string(),
+    title: truncateSchema({ maxCharacters: 70 }),
```
Comment (bautistaaa) [inline apps/og-image/utils/zodParams.ts:69]: i dont think you need to worry about og-image

you can move these zod utils to web

File: apps/web/src/app/[locale]/[username]/page.tsx (hunk 3 above)
Comment (bautistaaa) [reply apps/web/src/app/[locale]/[username]/page.tsx:14]: i dont believe github allows spaces in usernames so this should never happen

File: apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx [hunk 6]
```diff
@@ -82,29 +146,8 @@ export function CommentInput({ mode, onCancel, onChange, value, placeholder, onS
...
-
-                await onSubmit();
-              } catch (e) {
-                console.error(e);
-              } finally {
-                setIsSubmitting(false);
-                setCommentMode('editor');
-              }
-            }}
+            disabled={value.length === 0 || isSubmitting || !formValid}
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx:149]: this should just be type submit so it submits the form on click

File: apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx [hunk 7]
```diff
@@ -48,28 +67,73 @@ export function CommentInput({ mode, onCancel, onChange, value, placeholder, onS
     }
   };
 
+  const submitComment = async () => {
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx:70]: can this all be moved to a handleSubmit (see other forms in the app)

File: apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx [hunk 8]
```diff
@@ -32,6 +45,12 @@ export function CommentInput({ mode, onCancel, onChange, value, placeholder, onS
     }
 
     if (e.shiftKey && e.key === 'Enter') {
+      const success = singleFieldSchema.safeParse(form.getValues()).success;
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx:48]: is there a way to just trigger a form submit?

File: apps/web/src/components/wizard/default-tests.md [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// TEST CASE START (code in test cases are not editable)
+// TEST CASE START (code in test cases is not editable)
```
Comment (bautistaaa) [inline apps/web/src/components/wizard/default-tests.md:1]: this was correct already.

File: apps/web/src/utils/profanity.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import Filter from 'bad-words';
+import overrideWords from './bad-word-overrides.json';
+
+const f = new Filter();
+
+export function containsProfanity(str: string): boolean {
+  f.addWords(...overrideWords);
```
Comment (bautistaaa) [inline apps/web/src/utils/profanity.ts:7]: cant this be moved outside the functyion as well?

File: packages/ui/src/components/textarea.tsx [hunk 11]
```diff
@@ -6,14 +6,16 @@ export type TextareaProps = React.TextareaHTMLAttributes<HTMLTextAreaElement>;
...
     return (
-      <textarea
-        className={cn(
-          'border-input ring-offset-background placeholder:text-muted-foreground focus-visible:ring-ring flex min-h-[80px] w-full rounded-md border bg-transparent px-3 py-2 text-sm focus-visible:outline-none focus-visible:ring-2 disabled:cursor-not-allowed disabled:opacity-50',
-          className,
-        )}
-        ref={ref}
-        {...props}
-      />
+      <div>
```
Comment (bautistaaa) [inline packages/ui/src/components/textarea.tsx:9]: why did you add a wrapping div

**Review by bautistaaa — CHANGES_REQUESTED** [review]

File: packages/ui/src/components/textarea.tsx (hunk 11 above)
Comment (Rand0mLetters) [reply packages/ui/src/components/textarea.tsx:9]: oh I thought I removed that, but I guess I didn't. I'll fix that

File: apps/web/src/utils/profanity.ts (hunk 10 above)
Comment (Rand0mLetters) [reply apps/web/src/utils/profanity.ts:7]: yeah, I didn't notice that. will fix

File: apps/web/src/app/[locale]/[username]/page.tsx (hunk 3 above)
Comment (Rand0mLetters) [reply apps/web/src/app/[locale]/[username]/page.tsx:14]: yeah that's true, I just noticed the %20 in prod at /@Alexandra Osinski (there may be others as well) from the seed data and it bugged me

it won't happen with just github accounts, but it will with the seed data

should I remove it?

File: apps/og-image/utils/zodParams.ts (hunk 5 above)
Comment (Rand0mLetters) [reply apps/og-image/utils/zodParams.ts:69]: ok

File: apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx (hunk 8 above)
Comment (Rand0mLetters) [reply apps/web/src/app/[locale]/challenge/_components/comments/comment-input.tsx:48]: I'm not sure... it needs to return if the inputs aren't valid, so I think safeParse is necessary, right?

File: apps/web/src/app/[locale]/[username]/page.tsx (hunk 3 above)
Comment (Hacksore) [reply apps/web/src/app/[locale]/[username]/page.tsx:14]: you are good to decode here.

File: apps/og-image/utils/zodParams.ts (hunk 5 above)
Comment (Hacksore) [reply apps/og-image/utils/zodParams.ts:60]: remove all changes to og image pls

**Review by Hacksore — CHANGES_REQUESTED** [review]
Once og is reverted this is :shipit:'able

File: apps/og-image/utils/zodParams.ts (hunk 5 above)
Comment (Rand0mLetters) [reply apps/og-image/utils/zodParams.ts:60]: aight

**Comment by bautistaaa** [comment]
all the checks were failing. can you resolve so we can merge

**Review by bautistaaa — APPROVED** [review]
